common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java

```diff
@@ -0,0 +1,137 @@
+        projection.multiply(Vector3f.YP.rotationDegrees(camera.getYRot() + 180.0F));
+
+        // offset to put text to the center of the block
+        float dx = (float) (location.x + 0.5 - cameraPos.x);
+        float dy = (float) (location.y + 0.5 - cameraPos.y);
+        float dz = (float) (location.z + 0.5 - cameraPos.z);
+
+        double squaredDistance = dx * dx + dy * dy + dz * dz;
+
+        double distance = Math.sqrt(squaredDistance);
+        int maxDistance = McUtils.mc().options.renderDistance * 16;
+
+        this.distanceText = Math.round((float) distance) + "m";
+
+        // move the position to avoid ndc z leak past 1
+        if (distance > maxDistance) {
+            double posScale = maxDistance / distance;
+            dx *= posScale;
+            dz *= posScale;
+        }
+
+        worldToNdc(new Vector3f(dx, dy, dz), projection);
+    }
+
+    @SubscribeEvent
+    public void onRenderGuiPost(RenderEvent.Post event) {
+        if (CompassModel.getCompassLocation().isEmpty() || normalizedDeviceCoordinates == null || isOnScreen()) return;
+
+        Window window = event.getWindow();
+
+        float backgroundWidth = FontRenderer.getInstance().getFont().width(distanceText);
+        float backgroundHeight = FontRenderer.getInstance().getFont().lineHeight;
+        WynntilsMod.info(String.valueOf(backgroundWidth));
+
+        float displayPositionX = (float) ((normalizedDeviceCoordinates.x + 1.0f) / 2.0f) * window.getGuiScaledWidth();
+        float displayPositionY = (float) ((1.0f - normalizedDeviceCoordinates.y) / 2.0f) * window.getGuiScaledHeight();
+
+        RenderUtils.drawRect(
+                event.getPoseStack(),
+                CommonColors.BLACK.withAlpha(backgroundOpacity),
+                displayPositionX - (backgroundWidth / 2) - 2,
+                displayPositionY - (backgroundHeight / 2) - 2,
+                0,
+                backgroundWidth + 3,
+                backgroundHeight + 2);
+        FontRenderer.getInstance()
+                .renderAlignedTextInBox(
+                        event.getPoseStack(),
+                        distanceText,
+                        displayPositionX - backgroundWidth,
+                        displayPositionX + backgroundWidth,
+                        displayPositionY - backgroundHeight,
+                        displayPositionY + backgroundHeight,
+                        0,
+                        textColor,
+                        HorizontalAlignment.Center,
+                        VerticalAlignment.Middle,
+                        textShadow);
+    }
+
+    private boolean isOnScreen() {
+        return normalizedDeviceCoordinates.x <= 1
+                && normalizedDeviceCoordinates.x >= -1
+                && normalizedDeviceCoordinates.y <= 1
+                && normalizedDeviceCoordinates.y >= -1
+                && normalizedDeviceCoordinates.z > 1;
+    }
+
+    private void worldToNdc(Vector3f delta, Matrix4f projection) {
```

Comment:
This side-effect coding makes the function very opaque. Better if it returns a Vector3d, and you make the assignment at the call site.